Improved Docker Image

galexrt opened the pull request:
This pull request adds an improved `Dockerfile` and `docker-start.sh`.

The new `Dockerfile` removes the mysql server from it, because it isn't the docker way having more than "one" service running in the container. That's what `docker-compose.yml` is for (`docker-compose up`), to create a multi service application (staytus and mysql database server).

The `docker-start.sh` sets up the config file from the environment variables on every start, so you can move from server to server without the need to move the config file too. You would just run `docker-compose up` again to start. That's what Docker is thought for, easy setup (through `docker-compose`). Volumes are for persistent data, not really for config files. ;)

This improves the experience with staytus in combination with docker. :)

Currently the `docker-compose.yml` links to my image on the docker hub, but you or I can change it to fit your docker hub account.

If you have any questions regarding the new `Dockerfile`, just ask.

Thanks for creating Staytus!

astopy commented on Dockerfile:
```diff
@@ -3,26 +3,21 @@ MAINTAINER Tim Perry <pimterry@gmail.com>
+    apt-get clean && \
+    rm -rf /var/lib/apt/lists/* /tmp/* /var/tmp/* && \
+    mkdir -p /opt/staytus && \
+    git clone https://github.com/adamcooke/staytus.git /opt/staytus && \
```
It shouldn't be necessary to clone the repo here. Just `ADD . /opt/staytus`. (Cloning the repo like this will also make it impossible to build a docker image from any branch other than master.)

Galexrt commented:
@astopy I fixed it. It's now using the files from the current folder.

pimterry commented on Dockerfile:
```diff
@@ -3,26 +3,20 @@ MAINTAINER Tim Perry <pimterry@gmail.com>
+    # Instal node as the JS engine for uglifier
+    apt-get install -y nodejs && \
+    apt-get clean && \
+    rm -rf /var/lib/apt/lists/* /tmp/* /var/tmp/* && \
```
The `apt-get clean` here isn't necessary, and *maybe* some or all of the `rm` too (but I'm not sure there).

Automatic `apt-get clean` is baked into the Docker base images already, and gets run automatically for you: https://github.com/docker/docker/blob/03e2923e42446dbb830c654d0eec323a0b4ef02a/contrib/mkimage/debootstrap#L82-L105. You can see this in `/etc/apt/apt.conf.d/docker-clean` in the current image.

pimterry commented on Dockerfile:
```diff
@@ -3,26 +3,20 @@ MAINTAINER Tim Perry <pimterry@gmail.com>
-
-ENTRYPOINT /opt/staytus/docker-start.sh
+ADD docker-start.sh /docker-start.sh
+ADD . /opt/staytus/
```
Minor point, but COPY is preferable to ADD: https://docs.docker.com/engine/userguide/eng-image/dockerfile_best-practices/#add-or-copy

pimterry commented on Dockerfile:
```diff
@@ -3,26 +3,20 @@ MAINTAINER Tim Perry <pimterry@gmail.com>
+    apt-get clean && \
+    rm -rf /var/lib/apt/lists/* /tmp/* /var/tmp/* && \
+    cd /opt/staytus && \
+    bundle install --deployment --without development:test
```
Ordering is a interesting here.

In the original Dockerfile, all the generic setup (installing node, and the now-defunct mysql setup) was done in a separate preceding RUN step, before then COPYing and RUNing the parts that depend on content in the local directory.

Local development of the Dockerfile is much easier that way; you don't invalidate the old steps as you change content, so rebuilding the image is quite a bit quicker, and doesn't apt-get update from scratch and reinstall node again every time. This will work, but reorganising it just slightly makes local docker dev much easier.

pimterry commented on Dockerfile:
```diff
@@ -3,26 +3,20 @@ MAINTAINER Tim Perry <pimterry@gmail.com>
-    apt-get install -y mysql-server nodejs
-    
-COPY . /opt/staytus
+ENV DEBIAN_FRONTEND="noninteractive"
```
A bit debatable, but it's slightly nicer to keep this inside the RUN statement.

Having it in the RUN scopes it to just exist for that command (see http://stackoverflow.com/questions/33379393/docker-env-vs-run-export). Nothing else needs this var, and we certainly don't need to have it set in the resulting Docker image. Doesn't cost much obviously, but it's a bit tidier without.

pimterry commented on docker-compose.yml:
```diff
@@ -0,0 +1,18 @@
+  image: "sameersbn/mysql:latest"
+  environment:
+    DB_USER: staytus
+    DB_PASS: staytus
```
This means we've now got a hardcoded database username/password, which is slightly messy security-wise. It'd be nice if we could randomly generate a password and share it between the two here. I'm not sure how to do that nicely in docker-compose mind.

The goal here hopefully is to having something people can actually take a throw up into production to get a status site really easily & safely, without carefully configuring everything themselves. Having a production site running with totally default standard passwords like this isn't good.

Doesn't necessarily need to block this, since it's not an urgent problem, but secure-by-default would be a good thing to aim for if there's an easy route there.

Galexrt replied:
The database container is only reachable on the host system, not publicly on the internet. So this should only be a problem, if the webserver is breached and someone can access the host network.

I don't think `docker-compose` offers a way to generate a random password. But we could add a starting script to replace the password with a secure one and then start it for them, how about that?

pimterry replied:
Mmm, it shouldn't be directly exposed, but good security generally revolves around ensuring the whole tower of cards doesn't come tumbling down if any one thing fails. E.g. https://www.owasp.org/index.php/Defense_in_depth. It'd be good to be distributing software that's following security best practices, since people probably are going to put this straight into production. At the same time though, you're right that it's probably fine, and it's certainly not an actual vulnerability.

For now, if you can come up with a nice easy (i.e. no user interaction) way to generate a random password, that would be amazing. If nothing springs to mind though I think we should ignore this for now and create a new issue to think about doing it better later. It's not a big problem, more of a nice to have.

Maybe we could add a security note to the readme; if you're putting this into production, you should change the password below to minimize security risks? That'll probably do.

pimterry commented on docker-start.sh:
```diff
@@ -1,31 +1,25 @@
+sed -i "s|password:.*|password: \"$DB_PASSWORD\"|" /opt/staytus/config/database.yml
+sed -i "s|database:.*|database: $DB_DATABASE|" /opt/staytus/config/database.yml
 
-  bundle exec rake staytus:build staytus:install
```
Without this line we no longer run staytus:install, only staytus:upgrade (on L23 below).

Looks like that does actually work fine, which is slightly surprising, but we should watch out for this. If there's ever required first-run setup in staytus:install (very plausible), this might quietly break.

Given that staytus:upgrade seems to work fine for new instance setup here right now though, maybe we should just drop staytus:install entirely, and make upgrade into the only task available, handling the "initial install or update" case. @adamcooke?

pimterry commented on docker-start.sh:
```diff
@@ -1,31 +1,25 @@
-  # Use the previously saved config from the persisted volume
-  cp /opt/staytus/persisted/config/database.yml config/database.yml
-  # TODO also copy themes back and forth too
+cd /opt/staytus || exit 1
```
Why `|| exit 1` here? We've already copied things into /opt/staytus subdirectories above. Is this `cd` ever going to fail?

Galexrt replied:
@pimterry Yes you are right, this `cd` should never fail. I'm going to remove it.

pimterry commented:
Thanks @Galexrt, lots of great stuff here. I added the original Docker setup (https://github.com/adamcooke/staytus/pull/23 & https://github.com/adamcooke/staytus/pull/44), but docker-compose wasn't really recommended for production use at that point, and I never got around to going back and tidying this up.

These changes mostly look great, I've just sprinkled a few minor points and tweaks through in a couple of places. This will also need some doc updates to the [README Docker instructions](https://github.com/adamcooke/staytus#setup-with-docker), to tell users how to start it after these changes.

This probably ties nicely into https://github.com/adamcooke/staytus/issues/52 once it's done. If you briefly poke the settings on DockerHub @adamcooke you can get the image to rebuild automatically on changes to this project, and on updates (e.g. security patches) for the underlying Ruby image.

Galexrt commented:
@pimterry I look into adding the improvements in the upcoming week.
I'll change the README for Docker accordingly too.

Galexrt commented:
@pimterry I have added/changed the following things now:
* Added the improvements you suggested.
* Added a README.docker.md.
* Added the missing ports in the `docker-compose.yml`.
* Rebased all my commits.

One thing that we still need to discuss, is how we handle the password from the `docker-compose.yml`.

Galexrt commented:
It seems like I did something wrong during rebasing.. If you want me to create a new pr with only my commits.

pimterry commented on README.docker.md:
```diff
@@ -0,0 +1,32 @@
+## Setup with Docker
+
+1. [Install Docker locally](https://docs.docker.com/installation/)
+2. [Install docker-compose locally](https://docs.docker.com/installation/)
```
These URLs should be different. You probably want https://docs.docker.com/compose/install/

Galexrt replied:
Thanks, I fixed the link.

pimterry commented on README.docker.md:
```diff
@@ -0,0 +1,32 @@
+1. [Install Docker locally](https://docs.docker.com/installation/)
+2. [Install docker-compose locally](https://docs.docker.com/installation/)
+3. Now run `docker-compose -d up`, to start Staytus with the database container (You have to be in the directory of your Staytus `docker-compose.yml` for this).
+4. Go to [http://localhost:80](http://localhost:80) and follow the instructions to configure Staytus
```
I think this needs you to have a checkout of the codebase, or at least a local copy of the docker-compose.yml, right? One of the nice things with the current setup is that you don't have to do that; you just ask docker to run staytus, and then you're done. Looks like there isn't yet a neat solution to this in Docker (https://github.com/docker/compose/issues/1818), but an initial one's en route imminently: https://github.com/docker/compose/pull/3098.

For now, I think just add a step to clone the codebase locally, or a direct link to download the docker-compose here, so people know what's required.

As soon as that PR is merged and released though we should recommend just:

* Install docker
* Install docker-compose
* Run `docker-compose -f https://github.com/adamcooke/staytus/raw/master/docker-compose.yml up`

Galexrt replied:
I changed the run command to `curl https://github.com/adamcooke/staytus/blob/master/docker-compose.yml | docker-compose -f - up -d`, as someone suggested in the comment https://github.com/docker/compose/issues/1818#issuecomment-164807629.

pimterry replied:
Good spot! Very nice.

Galexrt commented:
@pimterry I added the improvements from your comments.

pimterry commented:
Great, thanks. All looks good to me :+1:

pimterry commented:
@adamcooke Any chance you can take a look at this? It's all ready now I think, it'd be good to get it merged in. Turning on #52 would be good too, so the Docker Hub build automagically updates on changes.

adamcooke commented:
I'm worried about the current integration with Docker and I would welcome and advise. I have no experience with it whatsoever so I am nervous at every pull request as I have no idea what the best way to go about these things is. I would love to be able to create a repository which contains just the Docker stuff which can be maintained by someone other than me who has experience in this area.

I'm happy to merge this (although it's currently got conflicts - sorry for the delay!) if there's no other alternative but having Docker specific configuration in the core repository doesn't sit well. What happens when there's an alternative to Docker, are we expected to maintain configuration or both in here?

galexrt commented:
@adamcooke I'm going to resolve the merge conflicts in the next one or two days.
I'm going to add some changes to the Docker container entrypoint so we don't need to make changes for the future, so we don't need to add every new setting to it. Would that be an idea?

adamcooke commented:
Sounds great - thank you

galexrt commented:
@adamcooke I have now pushed the more dynamic Docker entrypoint.
While creating the dynamic Docker entrypoint, I screwed up and deleted my Docker README file.

I'm going to add it in my rebased commit. It should be up in about two hours.

galexrt commented:
I just pushed the content of the `README.docker.md`.

I tested the new Docker image now, but I'm running into two errors when running `bundle exec foreman start`:
```
root@570cfddd0888:/opt/staytus# bundle exec foreman start    
08:57:10 web.1    | started with pid 13684
08:57:10 worker.1 | started with pid 13685
08:57:10 web.1    | /usr/local/lib/ruby/gems/2.3.0/gems/bundler-1.12.5/lib/bundler/rubygems_integration.rb:373:in `block in replace_bin_path': can't find executable bundle (Gem::Exception)
08:57:10 web.1    | 	from /usr/local/lib/ruby/site_ruby/2.3.0/rubygems.rb:298:in `activate_bin_path'
08:57:10 web.1    | 	from /usr/local/bundle/bin/bundle:22:in `<main>'
08:57:10 worker.1 | /usr/local/lib/ruby/gems/2.3.0/gems/bundler-1.12.5/lib/bundler/rubygems_integration.rb:373:in `block in replace_bin_path': can't find executable bundle (Gem::Exception)
08:57:10 worker.1 | 	from /usr/local/lib/ruby/site_ruby/2.3.0/rubygems.rb:298:in `activate_bin_path'
08:57:10 worker.1 | 	from /usr/local/bundle/bin/bundle:22:in `<main>'
08:57:10 web.1    | exited with code 1
08:57:10 system   | sending SIGTERM to all processes
08:57:10 worker.1 | exited with code 1
root@570cfddd0888:/opt/staytus#
```

Could you tell me what these errors mean, so I can fix them?

***

**Note:**
With the dynamic entrypoint, you now specify the settings using environment variables.
For example for the setting `STAYTUS_SMTP_HOSTNAME` the env var would be `SETTING_STAYTUS_SMTP_HOSTNAME`. (This even works with lower case vars, example `example_user_test` would be `SETTING_example_user_test`.

adamcooke commented:
Puzzling. Bundler is obviously installed because you're running `bundle exec foreman start`. I've not seen that before whenever I've run it. Perhaps foreman is doing something odd with the environment in docker?

galexrt commented:
I did some research and it seems that this is caused by a bug in bundler.. (see https://github.com/bundler/bundler/issues/4576).

galexrt commented:
@adamcooke I fixed the error by implementing a workaround mentioned in the issue https://github.com/bundler/bundler/issues/4576.

My changes are now ready to be merged. :)

pimterry commented on docker-compose.yml:
```diff
@@ -0,0 +1,22 @@
+    - "/opt/docker/staytus-database:/var/lib/mysql"
+staytus:
+  #image: "adamcooke/staytus:latest"
+  image: "staytus-test"
```
Is this correct? I think the commented out name is what we should be using for end-users - this requires you've got a staytus-test image already available locally.

galexrt replied:
Oops, yes you are right the outcommented is for the end users. I'm gonna fix this in a moment.

pimterry commented:
@adamcooke Thanks for taking a look at this (sorry it's taken so long for me to join in on this again too!)

Your concerns about having the Docker config here make sense, I think splitting them out is quite sensible, even with the more configurable setup here. Feels unnecessary to tie the distribution code to the larger Staytus codebase, splitting out reviewing would make life easier too, and it's then easier for us to manage setting up integration with Docker Hub etc.

If you want to do this, I'd be happy to maintain it (and I suspect @galexrt would too?). We should probably move this repo and the new one out into a standalone github organisation to do that though, rather than distributing these repos all over. 

How would you feel about that? Do you mind setting that up? Should just be a matter of creating the org, moving this repo there (which will automatically set up redirects from the current location for everything), creating a staytus-docker repo too, and adding the two of us to it.

galexrt commented:
@pimterry There should be no problem in just leaving the three (four with the docs) Docker related files in the repo. File list:
* `Dockerfile`
* `docker-compose.yml`
* `docker-entrypoint.sh`

The image entrypoint I created should be sufficient for the future as if you introduce new variables for example for the Staytus config variable `LDAP_HOST` Docker users just add `SETTING_LDAP_HOST` (with value) to their `docker-compose.yml` file and are good to go.
So the Docker image should have almost zero "maintenance" time. The only time I see time required to fix something with the image is when ground breaking gets changed like the installation procedure.
But even if a change like this occurs you can just hit me and @pimterry up and we'd be glad to fix it fast.

***

My opinion on moving Docker stuff to a new repo is that there is no need to move the Docker image to a separate repo as we talk about small 4 files and not hundreds of files that are required for the Docker image to be built.

adamcooke commented:
@pimterry I like the idea of not having any deployment specific things in the core repo so a new repo would be great. I'm not too worried about creating an org for it, perhaps you'd like to create a repository in your own account that we can just link to from our docs in this repo?

bersace commented:
@galexrt thanks for your contribution. You're doing it right. Shipping a docker image is not shipping a Vagrantfile or something packed. How to use another database service (say Amazon RDS or so) ?

galexrt commented:
@bersace It isn't documented in the Docker `README` but take a look at these lines in the entrypoint script: https://github.com/galexrt/staytus/blob/960abd8152746b4503d97b5cb24e54821fe96cc2/docker-entrypoint.sh#L8-L15

bersace commented:
@galexrt That is the right way :D

adamlc commented:
Would be great to see this merge and an automated docker build setup. Noticed the current docker hub image hasn't been updated for 2 years!

PeterDaveHello commented:
Not sure if this PR is still active? Would like to see it get merged, or if it's not active now, I can send another one :)

galexrt commented:
@PeterDaveHello I'm happy to update the PR, but I didn't get feedback yet on how to continue with the PR.

/cc @adamcooke

maddprof commented:
Is this PR full on dead at this point? The docker image in the repository is extremely dated.

galexrt commented:
@maddprof Well as written in the above comment, I have not gotten any feedback from @adamcooke (only @pimterry) over the span of creating the PR till now. I'm still hoping to get a review to update the PR.

@maddprof and anyone using my Docker image, if it is outdated, you can just give me a ping by creating an issue on the GitHub repo for it. I have gone ahead and updated the image, builds are complete right now.

galexrt commented:
Closing this PR. I don't see this being merged any time in the future.

adamcooke commented:
I do have a bunch of updating coming very soon to address exactly the contents of this PR. I already have a Dockerfile template that I use elsewhere that I'll probably use but it will remove the MySQL server etc.
